src/host_abi.zig

```diff
@@ -0,0 +1,124 @@
+///
+/// This is used in both calls from actual hosts as well as evaluation of constants
+/// inside the Roc compiler itself.
+pub fn RocOps(comptime CallEnv: type, comptime HostFns: type) type {
+    return struct {
+        /// The host provides this pointer, and Roc passes it to each of the following
+        /// function pointers as a second argument. This lets the host do things like use
+        /// arena allocators for allocation and deallocation (by putting the arena in here).
+        /// The pointer can be to absolutely anything the host likes, or null if unused.
+        env: *CallEnv,
+        /// Similar to  _aligned_malloc - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc
+        roc_alloc: fn (*RocAlloc, *CallEnv) void,
+        /// Similar to  _aligned_free - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-free
+        roc_dealloc: fn (*RocDealloc, *CallEnv) void,
+        /// Similar to  _aligned_realloc - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-realloc
+        roc_realloc: fn (*RocRealloc, *CallEnv) void,
+        /// Called when the Roc program has called `dbg` on something.
+        roc_dbg: fn (*RocDbg, *CallEnv) void,
+        /// Called when the Roc program has run an `expect` which failed.
+        roc_expect_failed: fn (*RocExpectFailed, *CallEnv) void,
+        /// Called when the Roc program crashes, e.g. due to integer overflow.
+        /// This function must not return, because the Roc program assumes it will
+        /// not continue to be executed after this function is called.
+        roc_crashed: fn (*RocCrashed, *CallEnv) void,
+        /// At the end of this struct, the host must include all the functions
+        /// it wants to provide to the Roc program for the Roc program to call
+        /// (e.g. I/O operations and such).
+        host_fns: HostFns,
+    };
+}
+
+/// When RocOps.roc_alloc gets called, it will be passed one of these.
+/// That function should write the address to the allocated memory into `ret`.
+/// If it cannot provide a non-null pointer (e.g. due to OOM), it
+/// must not return, and must instead do something along the lines
+/// of roc_crashed.
+pub const RocAlloc = struct {
+    alignment: usize,
+    length: usize,
+    answer: *anyopaque,
+};
+
+/// When RocOps.roc_dealloc gets called, it will be passed one of these.
+/// (The length of the allocation cannot be provided, because it is
+/// not always known at runtime due to the way seamless slices work.)
+pub const RocDealloc = struct {
+    alignment: usize,
+    ptr: *anyopaque,
+};
+
+/// When RocOps.roc_realloc gets called, it will be passed one of these.
+/// That function should write the allocated memory into `ret`.
+/// If it cannot provide a non-null pointer (e.g. due to OOM), it
+/// must not return, and must instead do something along the lines
+/// of roc_crashed.
+pub const RocRealloc = struct {
```

Comment:
I don't understand the full context for this PR but if you want this type to have well-defined memory layout, I think you want this:

```suggestion
pub const RocRealloc = extern struct {
```

This limits what types the struct can have inside of it and makes it adhere to the C ABI for the target.